### ISSUE TITLE
☂️ linter: eslint-recommended rules

### ISSUE THREAD

**Issue by Boshen** [Boshen/oxc#479]
These are the rules that we need to implement, except the ones that are:
* stylistic
* conventional
* strict mode syntax errors

https://github.com/eslint/eslint/blob/main/packages/js/src/configs/eslint-recommended.js

- [x] constructor-super
- [x] for-direction
- [x] getter-return
- [x] no-async-promise-executor
- [ ] no-case-declarations
- [x] no-class-assign
- [x] no-compare-neg-zero
- [ ] no-cond-assign
- [x] no-const-assign
- [x] no-constant-condition
- [ ] no-control-regex
- [x] no-debugger
- [ ] no-delete-var
- [ ] no-dupe-args
- [x] no-dupe-class-members
- [ ] no-dupe-else-if
- [x] no-dupe-keys
- [x] no-duplicate-case
- [ ] no-empty
- [ ] no-empty-character-class
- [x] no-empty-pattern
- [ ] no-ex-assign
- [ ] no-extra-boolean-cast
- [ ] no-extra-semi
- [ ] no-fallthrough
- [x] no-func-assign
- [ ] no-global-assign
- [ ] no-import-assign
- [ ] no-inner-declarations
- [ ] no-invalid-regexp
- [ ] no-irregular-whitespace
- [ ] no-loss-of-precision
- [ ] no-misleading-character-class
- [ ] no-mixed-spaces-and-tabs
- [x] no-new-symbol
- [ ] no-nonoctal-decimal-escape
- [ ] no-obj-calls
- [ ] no-octal (strict mode syntax error)
- [ ] no-prototype-builtins
- [ ] no-redeclare
- [ ] no-regex-spaces
- [ ] no-self-assign
- [x] no-setter-return
- [ ] no-shadow-restricted-names
- [ ] no-sparse-arrays
- [ ] no-this-before-super
- [ ] no-undef
- [ ] no-unexpected-multiline
- [ ] no-unreachable
- [ ] no-unsafe-finally
- [x] no-unsafe-negation
- [ ] no-unsafe-optional-chaining
- [x] no-unused-labels
- [ ] no-unused-vars
- [ ] no-useless-backreference
- [ ] no-useless-catch
- [ ] no-useless-escape
- [ ] no-with (strict mode syntax error)
- [ ] require-yield
- [x] use-isnan
- [x] valid-typeof

**Comment by DonIsaac** [comment]
@Boshen can we please update this?

**Comment by Boshen** [comment]
> @Boshen can we please update this?

Feel free to edit it, you have permission to do so.

**Comment by suica** [comment]
For new comers:
1. Use `just new-rule <name>` to generate a rule file with test cases, as mentioned in https://github.com/web-infra-dev/oxc/issues/597. The `just new-rule <name>` command will fetch the info of eslint rule with exact the same name and fill the template for you.
1. Remember to declare the rule to `crates/oxc_linter/src/rules.rs`, to make sure the rule file is in the file hierarchy, otherwise the language server will not work in generated rule file, and VSCode will also complain about it.
1. Start to implement the rule and have fun!

**Comment by nrayburn-tech** [comment]
unicorn/no-nested-ternary has been implemented.  Should eslint/no-nested-ternary be marked as complete? Or something else done for this rule?

The eslint version doesn't allow any nesting.  The unicorn version allows nesting 1 level, as long as it is wrapped in parenthesis.
https://eslint.org/docs/latest/rules/no-nested-ternary
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/rules/no-nested-ternary.md

**Comment by connorshea** [comment]
Given #7342, should `no-restricted-syntax` be added as unsupported?

**Comment by camc314** [comment]
> Given #7342, should `no-restricted-syntax` be added as unsupported?

I don't think so, the reason being is that we are going to have too support esquery like syntax, in order to achieve good performance for custom JS plugins, but i think it'll take a while to get to.

**Comment by connorshea** [comment]
Deprecated rules, especially ones like https://eslint.org/docs/latest/rules/indent-legacy, should probably be considered for marking as 'not needed' here, right?

**Comment by camc314** [comment]
> Deprecated rules, especially ones like https://eslint.org/docs/latest/rules/indent-legacy, should probably be considered for marking as 'not needed' here, right?

Yes. i'll update the task,

**Comment by connorshea** [comment]
If `consistent-return` is going to be [implemented in tsgolint](https://github.com/oxc-project/tsgolint/issues/185) (assuming it is, maybe it will not be as it's duplicative of TypeScript's compiler checks), should it be marked as not-needed for ESLint core?

Should consider this for all the [extension rules](https://typescript-eslint.io/rules?=extension#rules).

**Comment by JSteunou** [comment]
I noticed that `object-shorthand` is listed here but not in the [rules documentation](https://oxc.rs/docs/guide/usage/linter/rules.html)

update: just seen the MR merged a few days ago, maybe just not published yet, sorry for the false alert if it is the case.

**Comment by camc314** [comment]
> I noticed that `object-shorthand` is listed here but not in the [rules documentation](https://oxc.rs/docs/guide/usage/linter/rules.html)
> 
> update: just seen the MR merged a few days ago, maybe just not published yet, sorry for the false alert if it is the case.

@JSteunou shoukd be released on monday ish.
